src/analysis/ml.py

```diff
@@ -152,27 +175,92 @@ def plot_dendrogram(model, **kwargs):
     # Plot the corresponding dendrogram
     dendrogram(linkage_matrix, **kwargs)
 
-
-def hac(dataframe: pd.DataFrame, output_png: str, output_file: str):
+def hac_vertical(dataframe: pd.DataFrame, output_png: str, output_file: str, linkage='ward', metric='euclidean'):
     """
     Performs hierarchical agglomerative clustering on the dataframe,
-    creates a dendrogram of the resulting tree,
+    creates a dendrogram of the resulting tree USING SEABORN and SCKIT-LEARN for the cluster groups,
     and saves the dendrogram and the cluster labels of said dendrogram in separate files.
     @param dataframe: binary dataframe of edge comparison between algorithms from summarize_networks
     @param output_png: the file name to save the dendrogram image
     @param output_file: the file name to save the clustering labels
     """
+
+    if linkage not in linkage_methods:
+        raise ValueError(f"linkage={linkage} must be one of {linkage_methods}")
+    if linkage == "ward":
+        if metric != "euclidean":
+            print("For linkage='ward', the metric must be 'euclidean'; setting metric = 'euclidean")
+            metric = "euclidean"
+    if metric not in distance_metrics:
+        raise ValueError(f"metric={metric} must be one of {distance_metrics}")
+        
     X = dataframe.reset_index(drop=True)
-    X = X.transpose()
-    model = AgglomerativeClustering(distance_threshold=0.5, n_clusters=None)
+    column_names = X.head()
+    column_names = [element.split('-')[1] for element in column_names]
+    X = X.transpose() 
+
+    # creating the colors per algorithms
+    custom_palette = sns.color_palette("tab10", len(column_names))
+    label_color_map = {label: color for label, color in zip(column_names, custom_palette)}
```

Comment:
Are we able to pass this colormap to the `sns.scatterplot` used for the PCA plot? If we can, then it would make the results consistent to define the colormap ourselves both times to make sure they are consistent. In that case, we could have a new small function in this file that takes a list of column names and returns the map in a deterministic way such as sorting the names.